src/main/java/org/jenkinsci/test/acceptance/docker/DockerContainerHolder.java

```diff
@@ -22,15 +25,35 @@
     @Inject
     Docker docker;
 
+    @Inject
+    private FailureDiagnostics diag;
+
     T container;
 
     /**
+     * Injecting a portOffset will force the binding of dockerPorts to local Ports with an offset
+     * (e.g. bind docker 22 to localhost port 40022,
+     */
+    @Inject(optional = true)
+    @Named("dockerPortOffset")
+    private int portOffset = 0;
+
+    /**
      * Lazily starts a container and returns the instance.
      */
     @Override
     public synchronized T get() {
-        if (container==null)
-            container = docker.start((Class<T>)type.getRawType());
+        if (container==null) {
+            Class<T> fixture = (Class<T>) type.getRawType();
+            File buildlog = diag.touch("docker-" + fixture.getSimpleName() + ".build.log");
+            File runlog = diag.touch("docker-" + fixture.getSimpleName() + ".run.log");
+            try {
+                System.out.println("Setting up docker container for " + fixture.getSimpleName());
+                container = docker.build(fixture, buildlog).start(fixture).withPortOffset(portOffset).withLog(runlog).start();
+            } catch (InterruptedException | IOException e) {
+                throw new AssertionError("Failed to start container " + fixture.getName(), e);
```

Comment:
if the `docker.build(fixture, buildlog)` fails then it throws an Error, which would not be translated into an AssertionError.

I think this is correct as I would want the test to be in `Error` not `Failed` - so I am wondering why we treat any error at all here as `AssertionError` rather than a different type of `Error` that would cause the test to be marked as `error` rather than `failed`?